posts/2019-04-11-Rust-1.34.0.md

```diff
@@ -0,0 +1,219 @@
+
+### `?` in documentation tests
+
+[RFC 1937]: https://rust-lang.github.io/rfcs/1937-ques-in-main.html
+[many releases ago]: https://blog.rust-lang.org/2018/05/10/Rust-1.26.html#main-can-return-a-result
+
+[RFC 1937] proposed adding support for using the `?` operator in `fn main()`,
+`#[test]` functions, and doctests, allowing them to return `Option<T>` or `Result<T, E>`,
+with error values causing a nonzero exit code in the case of `fn main()`,
+and a test failure in the case of the tests.
+
+Support in `fn main()` and `#[test]` was implemented [many releases ago].
+However, the support within documentation tests was limited to doctests that have an explicit `fn main()`.
+
+In this release, full support for `?` in doctests has been added.
+Now, you can write this in your documentation tests:
+
+````rust
+/// ```rust
+/// use std::io;
+/// let mut input = String::new();
+/// io::stdin().read_line(&mut input)?;
+/// # Ok::<(), io:Error>(())
+/// ```
+fn my_func() {}
+````
+
+You still have to specify the error type being used at the bottom of the documentation test.
+
+### Custom attributes accept arbitrary token streams
+
+[Procedural macros]: https://blog.rust-lang.org/2018/12/21/Procedural-Macros-in-Rust-2018.html
+[arbitrary-tts]: https://github.com/rust-lang/rust/pull/57367
+
+[Procedural macros] in Rust can define custom attributes that they consume.
+Until now, such attributes were restricted to being trees of paths and literals
+according to a specific syntax, like:
+
+```rust
+#[foo(bar)]
+#[foo = "bar"]
+#[foo = 0]
+#[foo(bar = true)]
+#[foo(bar, baz(quux, foo = "bar"))]
+```
+
+Unlike procedural macros, these helper attributes could not accept arbitrary token streams in delimiters,
+so you could not write `#[range(0..10)]` or `#[bound(T: MyTrait)]`.
+Procedural macro crates would instead use strings for specifying syntaxes like this, e.g. `#[range("0..10")]`
+
+With this Rust release, custom attributes `#[attr($tokens)]` [now accept][arbitrary-tts]
+arbitrary token streams in `$tokens`, bringing them on par with macros.
+If you're the author of a procedural macro crate, please check if your custom attributes
+have unnecessary strings in their syntax and if they can be better expressed with token streams.
+
+### `TryFrom` and `TryInto`
+
+[`from_be_bytes`]: https://doc.rust-lang.org/std/primitive.u32.html#method.from_be_bytes
+[never_type]: https://github.com/rust-lang/rust/issues/35121
+[`TryFrom`]: https://doc.rust-lang.org/std/convert/trait.TryFrom.html
+[`TryInto`]: https://doc.rust-lang.org/std/convert/trait.TryInto.html
+[`Infallible`]: https://doc.rust-lang.org/std/convert/enum.Infallible.html
+
+The [`TryFrom`] and [`TryInto`] traits were stabilized to allow fallible type conversions.
+
+For example, the [`from_be_bytes`] and related methods on integer types take arrays,
+but data is often read in via slices. Converting between slices and arrays is tedious to do manually.
```

Comment:
suggestion: ~~is~~ has been tedious

it no longer is 😃

Reply:
It is still tedious to do manually ;)